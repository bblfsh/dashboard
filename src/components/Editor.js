import React, { Component } from 'react';
import styled, { injectGlobal } from 'styled-components';
import CodeMirror from 'react-codemirror';

import 'codemirror/lib/codemirror.css';
import 'codemirror/theme/solarized.css';
import 'codemirror/mode/python/python';
import 'codemirror/mode/clike/clike';

// eslint-disable-next-line
injectGlobal`
  .ReactCodeMirror, .CodeMirror {
    height: 100%;
  }
`

const Container = styled.div`
  height: 100%;
`

export default class Editor extends Component {
  get document() {
    return this.refs.codemirror.getCodeMirror().getDoc();
  }

  selectCode(from, to) {
    return this.document.markText(from, to, {
      css: 'background: yellow',
    });
  }

  onCursorActivity(editor) {
    if (!this.props.onCursorChanged) {
      return;
    }

    const doc = editor.getDoc();
    const selection = doc.listSelections().slice(0, 1).pop();
    const positions = [selection.head, selection.anchor];
    positions.sort(comparePos);
    const [ start, end ] = positions;
    this.props.onCursorChanged({ start, end });
  }

  render() {
    const {
      code,
      languageMode,
      onChange,
    } = this.props;

    const options = {
      mode: languageMode,
      lineNumbers: true,
      theme: 'solarized light',
    };

    return (
      <Container>
        <CodeMirror 
          ref='codemirror'
          autoFocus={true}
          onChange={onChange}
          value={code}
          onCursorActivity={editor => this.onCursorActivity(editor)}
          options={options} />
      </Container>
    );
  }
}

export function comparePos(a, b) {
  if (a.line === b.line) {
    return a.ch - b.ch;
  }
  return a.line - b.line;
}
