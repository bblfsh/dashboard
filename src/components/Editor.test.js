import React from 'react';
import { shallow } from 'enzyme';
import ReactDOM from 'react-dom';
import toJson from 'enzyme-to-json';
import Editor, { comparePos } from './Editor';

const mkPos = (line, ch) => ({ line, ch });

describe('Editor', () => {
  it('renders without crashing', () => {
    const wrapper = shallow(
      <Editor code='foo = 1' languageMode='python' />
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('calls onCursorChanged when the selection changes', () => {
    const spy = jest.fn();
    const elem = document.createElement('div');
    const component = ReactDOM.render(
      <Editor code='foo = 1' languageMode='python' onCursorChanged={spy} />,
      elem,
    );

    component.document.setSelection(
      mkPos(0, 4),      
      mkPos(0, 2),
    );

    // NOTE: there is an initial call when the CodeMirror is rendered.
    expect(spy.mock.calls.length).toBe(2);
    const selection = spy.mock.calls[1][0];
    expect(selection).toEqual({ start: mkPos(0, 2), end: mkPos(0, 4) });
  });

  it('calls onCursorChanged when the cursor position changes', () => {
    const spy = jest.fn();
    const elem = document.createElement('div');
    const component = ReactDOM.render(
      <Editor code='foo = 1' languageMode='python' onCursorChanged={spy} />,
      elem,
    );

    component.document.setCursor(mkPos(0, 4));

    // NOTE: there is an initial call when the CodeMirror is rendered.
    expect(spy.mock.calls.length).toBe(2);
    const selection = spy.mock.calls[1][0];
    expect(selection).toEqual({ start: mkPos(0, 4), end: mkPos(0, 4) });
  });

  it('calls onChange when the content changes', () => {
    const spy = jest.fn();
    const elem = document.createElement('div');
    const component = ReactDOM.render(
      <Editor code='foo = 1' languageMode='python' onChange={spy} />,
      elem,
    );

    component.document.replaceRange('changed', mkPos(0, 0), mkPos(0, 3));

    expect(spy.mock.calls.length).toBe(1);
    expect(spy.mock.calls[0][0]).toBe('changed = 1');
  });

  it('selects the code when selectCode is called', () => {
    const elem = document.createElement('div');
    const component = ReactDOM.render(
      <Editor code='foo = 1' languageMode='python' />,
      elem,
    );

    const assertMarks = n => expect(component.document.getAllMarks().length).toBe(n);

    assertMarks(0);

    const mark = component.selectCode(mkPos(0, 0), mkPos(0, 4));
    assertMarks(1);

    mark.clear();
    assertMarks(0);
  });
});

describe('comparePos', () => {
  it('should correctly compare positions', () => {
    const cases = [
      { a: mkPos(1, 1), b: mkPos(1, 1), expected: 0 },
      { a: mkPos(1, 2), b: mkPos(1, 1), expected: 1 },
      { a: mkPos(1, 1), b: mkPos(1, 2), expected: -1 },
      { a: mkPos(2, 1), b: mkPos(1, 1), expected: 1 },
      { a: mkPos(1, 1), b: mkPos(2, 1), expected: -1 },
    ];

    cases.forEach(c => {
      expect(comparePos(c.a, c.b)).toBe(c.expected);
    });
  });
});
