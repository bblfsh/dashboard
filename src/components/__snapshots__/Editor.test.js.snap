// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Editor renders without crashing 1`] = `
<styled.div>
  <Component
    autoFocus={true}
    onCursorActivity={[Function]}
    options={
      Object {
        "lineNumbers": true,
        "mode": "python",
        "theme": "solarized light",
      }
    }
    preserveScrollPosition={false}
    value="foo = 1"
  />
</styled.div>
`;
